bindings/rust/s2n-tls/src/cert_chain.rs

```diff
@@ -75,20 +107,115 @@ impl CertificateChain<'_> {
+                certificate.as_ptr(),
+                private_key.as_ptr(),
+            )
+            .into_result()
+        }?;
+        Ok(self)
+    }
+
+    pub fn set_ocsp_data(&mut self, data: &[u8]) -> Result<&mut Self, Error> {
+        let size: u32 = data.len().try_into().map_err(|_| Error::INVALID_INPUT)?;
+        unsafe {
+            s2n_cert_chain_and_key_set_ocsp_data(self.ptr.as_ptr(), data.as_ptr(), size)
+                .into_result()
+        }?;
+        Ok(self)
+    }
+}
+
+impl Clone for CertificateChain<'_> {
+    fn clone(&self) -> Self {
+        let context = self.context();
+
+        // Safety
+        //
+        // Using a relaxed ordering is alright here, as knowledge of the
+        // original reference prevents other threads from erroneously deleting
+        // the object.
+        // https://github.com/rust-lang/rust/blob/e012a191d768adeda1ee36a99ef8b92d51920154/library/alloc/src/sync.rs#L1329
+        let _count = context.refcount.fetch_add(1, Ordering::Relaxed);
+        Self {
+            ptr: self.ptr,
+            is_owned: true, // clone only makes sense for owned
+            _lifetime: PhantomData,
+        }
+    }
 }
 
 // # Safety
 //
 // s2n_cert_chain_and_key objects can be sent across threads.
 unsafe impl Send for CertificateChain<'_> {}
 
+/// # Safety
+///
+/// Safety: All C methods that mutate the s2n_cert_chain are wrapped
+/// in Rust methods that require a mutable reference.
+unsafe impl Sync for CertificateChain<'_> {}
+
 impl Drop for CertificateChain<'_> {
     fn drop(&mut self) {
-        if self.is_owned {
-            // ignore failures since there's not much we can do about it
-            unsafe {
-                let _ = s2n_cert_chain_and_key_free(self.ptr.as_ptr()).into_result();
-            }
+        if !self.is_owned {
+            // not ours to cleanup
+            return;
+        }
+        let context = self.context_mut();
+        let count = context.refcount.fetch_sub(1, Ordering::Release);
+        debug_assert!(count > 0, "refcount should not drop below 1 instance");
+
+        // only free the cert if this is the last instance
+        if count != 1 {
+            return;
+        }
+
+        // Safety
+        //
+        // The use of Ordering and fence mirrors the `Arc` implementation in
+        // the standard library.
+        //
+        // This fence is needed to prevent reordering of use of the data and
+        // deletion of the data.  Because it is marked `Release`, the decreasing
+        // of the reference count synchronizes with this `Acquire` fence. This
+        // means that use of the data happens before decreasing the reference
+        // count, which happens before this fence, which happens before the
+        // deletion of the data.
+        // https://github.com/rust-lang/rust/blob/e012a191d768adeda1ee36a99ef8b92d51920154/library/alloc/src/sync.rs#L1637
+        std::sync::atomic::fence(Ordering::Acquire);
```

Comment:
I'm not sure if this fence is actually required. I followed drop from Config here.